lib/Bridge/TolerantParser/Reflector/TolerantSourceCodeReflector.php

```diff
@@ -48,7 +48,7 @@ public function reflectClassesIn($sourceCode): ReflectionClassCollection
     /**
      * {@inheritDoc}
      */
-    public function reflectOffset($sourceCode, $offset): ReflectionOffset
+    public function reflectOffset($sourceCode, $offset, $closestParent = false): ReflectionOffset
```

Comment:
I'm a bit nervous about adding behavioral flags, just for discussion (not diving into this right now), would it be possible do you think to acheive this without the flag? maybe with a way to retrieve the closest parent at a given offset?

Reply:
Well, if I remember correctly the thing is that it's not easy to identify if a symbol could be resolved to a node or not.
The offset is "reflected" which provide a symbol which is then resolved to a node.
That's this process that return an unknown node.
I tough the AST was a one way tree but if we can access the parent maybe.
The question is which guarantee do we have that the parent node can be resolved ?
The node resolver test the node against a bunch of different possible types and that's why I implemented it that way.
So the node resolver was the only one with the responsibility to resolve a symbol to node.
But we could try to reflect the offset of the parent of the unresolved node, then resolved this new symbol and do it again until we got a valid node or the root of the tree.
I will try to find some time to check it out this week-end, but no promises.

Reply:
So I took a look and:
I think it should be handled inside the code reflector, so it can be reused.
But we could get rid of the flag which may be an indicator that the method has too many responsibilities.
In which case I propose to provide another method to the `SourceCodeReflector` interface: `reflectOffsetToClosestParent`.
This method will have the same signature at the one currently used, but will try to reflect each node climbing up the AST tree until a resolvable node was found.

I already have a working version, I'm just finishing with the tests, here is the code in the mid time:
```php
    public function reflectOffsetToClosestParent($sourceCode, $offset): ReflectionOffset
    {
        $node = $this->nodeFromSourceAndOffset($sourceCode, $offset);

        $resolver = $this->serviceLocator->symbolContextResolver();
        $frame = $this->serviceLocator->frameBuilder()->build($node);

        do {
            $symbolContext = $resolver->resolveNode($frame, $node);

            $isUnknown = Symbol::UNKNOWN === $symbolContext->symbol()->symbolType();
        } while ($isUnknown && $node = $node->getParent());

        return TolerantReflectionOffset::fromFrameAndSymbolContext($frame, $symbolContext);
    }
```

Reply:
I am still not 100% convinced -- I don't think reflection API needs more methods necessarily, as this could also be done in a utility class. 

Also - the closet _resolvable_ parent depends on what this library supports, and so could produce unexpected results - so might be better to just resolve self or parent, make sensible decisions (if on whitespace, goto parent).

I'd like to suggest something else, as we have already seen some duplication in the `code-transform` and RPC libraries, maybe we add some utility methods to code-transform instead and support this, reduce duplication and decouple the RPC handler from WR?

For this case we could have a method to `getClosestResolvableOffset`, which can have a tolerant parser implementation which parses the AST (it's cached anyway) and walks up from the offset, trying to resolve the offset at each turn, finally returning an offset that can be resolved.

I can also have a look at that, hopefully this weekend.

Reply:
Well, you have a better vision of the global project and what every piece can/should do.
I'm working with big gaps from my part so I trust you on those decisions.

> the closet _resolvable_ parent depends on what this library supports, and so could produce unexpected results

Should not all the implementation be able to handle all the symbols ? No matter what kind of "node" or other thing they use ?
I saw `reflectOffsetToClosestParent` without thinking about an AST or nodes or anything, it's just mean that it should resolve to the closest understandable "thing" phpactor can use.

The main goal was to not deal with it in the handler, but it might not be the right place never less.

> maybe we add some utility methods to code-transform instead and support this

Again, I will trust you on where to put it. I just thought that `code-transform` will only handle code transformation. Here it's not really about changing the code, just being able to understand it.
Or is `code-transform`also including transforming the code into an understandable entity ?

To be honest I'm kind of lost between what WR does vs what the Microsoft Tolerant Parser does.
I didn't take the time to think about it yet, I was just throwing down code trying to mimic what I saw.

I agree with you, there is a subject to refactor some behaviour in the right place, but I also think it can be really helpful to access the context menu from anywhere.
Would it be possible to open an issue for the refactoring and still merge this solution in the mid while ?
Maybe already mark the new method as deprecated or something to not forget that it shouldn't be use ?

I won't have much time this week-end on my side, so I won't begin to work on that until maybe next week.

Reply:
> Should not all the implementation be able to handle all the symbols ? No
> matter what kind of "node" or other thing they use ?

Well, to be honest what a ReflectionOffset actually is is a bit vague, I
suppose really it could show all the possible information that can be inferred
from a specific offset in the code. At the moment only certain parts of the AST
are supported (and mostly these are enough). But we could f.e. return a
ReflectionOffset for, for example, whitespace, or a docblock (or an `if`
keyword, etc.) - none of which are currently supported, but _could_ be in the
future, so this method's behavior might change. (as far as I can tell).

> I just thought that code-transform will only handle code transformation

It handles code transformations/refactorings, but you could say that "locate
nearest thing I can refactor" is a question that library could answer. Indeed
-- I think it might even be a place for better context menu logic (f.e.
refactorings could indicate if they could be applied or not).

This is something I have been thinking of doing for a while now, and this might
be a good candidate to start with.

> Would it be possible to open an issue for the refactoring and still merge this solution in the mid while ?

I'd rather not -- but I will have a look at creating the helper method this afternoon.